src/main/scheduler/scheduler.c

```diff
@@ -32,247 +33,301 @@
+    if ((taskId == TASK_SELF && currentTask != NULL) || (taskId < (int)taskCount && taskId >= 0)) {
         cfTask_t *task = taskId == TASK_SELF ? currentTask : &cfTasks[taskId];
-        if (enabled && task->taskFunc) {
-            queueAdd(task);
-        } else {
-            queueRemove(task);
-        }
+        task->isEnabled = (enabled && task->taskFunc);
     }
 }
 
-uint32_t getTaskDeltaTime(const int taskId)
+uint32_t getTaskTimeSinceLastRun(const int taskId)
 {
-    if (taskId == TASK_SELF || taskId < (int)taskCount) {
+    if ((taskId == TASK_SELF && currentTask != NULL) || (taskId < (int)taskCount && taskId >= 0)) {
+        // Return the amount of time since the task last ran.
         cfTask_t *task = taskId == TASK_SELF ? currentTask : &cfTasks[taskId];
-        return task->taskLatestDeltaTime;
+        return currentTime - task->lastExecutionTime;
     } else {
         return 0;
     }
 }
 
 void schedulerInit(void)
 {
-    queueClear();
+    // Disable all tasks and set defaults
+    for(uint16_t ii = 0; ii < taskCount; ii++)
+    {
+        cfTasks[ii].isEnabled = false;
+        cfTasks[ii].isWaitingToBeRan = false;
+        cfTasks[ii].lastIdealExecutionTime = 0;
+        cfTasks[ii].averageExecutionTime = 0;
+        cfTasks[ii].lastExecutionTime = 0;
+    }
 }
 
-void scheduler(void)
+void schedulerExecute(void)
 {
     // Cache currentTime
     currentTime = micros();
 
-    // Check for realtime tasks
+    // Realtime Guard; We want to ensure whatever task we schedule now wont overrun a realtime task time.
+    // Find when the next realtime task will run.
     uint32_t timeToNextRealtimeTask = UINT32_MAX;
-    for (const cfTask_t *task = queueFirst(); task != NULL && task->staticPriority >= TASK_PRIORITY_REALTIME; task = queueNext()) {
-        const uint32_t nextExecuteAt = task->lastExecutedAt + task->desiredPeriod;
-        if ((int32_t)(currentTime - nextExecuteAt) >= 0) {
-            timeToNextRealtimeTask = 0;
-        } else {
-            const uint32_t newTimeInterval = nextExecuteAt - currentTime;
-            timeToNextRealtimeTask = MIN(timeToNextRealtimeTask, newTimeInterval);
+    for (uint16_t ii = 0; ii < taskCount; ii++) {
+        cfTask_t *task = &cfTasks[ii];
+        if(task->isEnabled && task->priority >= TASK_PRIORITY_REALTIME) {
+            const uint32_t nextExecuteAt = task->lastIdealExecutionTime + task->desiredPeriod;
+            if ((int32_t)(currentTime - nextExecuteAt) >= 0) {
+                timeToNextRealtimeTask = 0;
+            } else {
+                const uint32_t newTimeInterval = nextExecuteAt - currentTime;
+                timeToNextRealtimeTask = MIN(timeToNextRealtimeTask, newTimeInterval);
+            }
         }
     }
-    const bool outsideRealtimeGuardInterval = (timeToNextRealtimeTask > realtimeGuardInterval);
+
+#ifdef SCHEDULER_DEBUG_PRINT
+    printf("Time Until Next Realtime (%u)\n", timeToNextRealtimeTask);
```

Comment:
perhaps we should replace SCHEDULER_DEBUG_PRINT with DEBUG_PRINT(SCHEDULER, ...) macros or something?  not critical right now but I can see it being useful.